PlayFabSDK/source/Qos/PlayFabQosApi.cs

```diff
@@ -80,30 +98,56 @@ private async Task InitializeQoSServerList()
                     continue;
                 }
 
-                _dataCenterMap[qosServer.Region] = qosServer.ServerUrl;
+                dataCenterMap[qosServer.Region] = qosServer.ServerUrl;
             }
+
+            return _dataCenterMap = dataCenterMap;
         }
 
-        private async Task<List<QosRegionResult>> GetSortedRegionLatencies(int timeoutMs)
+        private async Task<QosResult> GetSortedRegionLatencies(int timeoutMs,
+            Dictionary<string, string> dataCenterMap, int pingsPerRegion, int degreeOfParallelism)
         {
-            var asyncPingResults = new List<Task<QosRegionResult>>(_dataCenterMap.Count);
-            foreach (KeyValuePair<string, string> datacenter in _dataCenterMap)
-            {
-                var regionPinger = new RegionPinger(datacenter.Value, datacenter.Key);
-                Task<QosRegionResult> pingResult = regionPinger.PingAsync(timeoutMs);
-                asyncPingResults.Add(pingResult);
-            }
+            RegionPinger[] regionPingers = dataCenterMap.Select(
+                datacenter => new RegionPinger(datacenter.Value, datacenter.Key, timeoutMs, NumTimeoutsForError, pingsPerRegion)).ToArray();
 
-            await Task.WhenAll(asyncPingResults);
-            var results = new List<QosRegionResult>(asyncPingResults.Count);
-            foreach (Task<QosRegionResult> asyncPingResult in asyncPingResults)
-            {
-                results.Add(asyncPingResult.Result);
-            }
+            ConcurrentBag<int> seeds = new ConcurrentBag<int>(Enumerable.Range(0, pingsPerRegion)
+                .Select(i => i * dataCenterMap.Count / pingsPerRegion));
+
+            Task[] pingWorkers = Enumerable.Range(0, degreeOfParallelism).Select(
+                i => PingWorker(regionPingers, seeds)).ToArray();
 
+            await Task.WhenAll(pingWorkers);
+
+            List<QosRegionResult> results = regionPingers.Select(x => x.GetResult()).ToList();
             results.Sort((x, y) => x.LatencyMs.CompareTo(y.LatencyMs));
 
-            return results;
+            QosErrorCode resultCode = QosErrorCode.Success;
+            string errorMessage = null;
+            if (results.All(x => x.ErrorCode == (int) QosErrorCode.NoResult))
+            {
+                resultCode = QosErrorCode.NoResult;
+                errorMessage = "No valid results from any QoS server";
+            }
+            
+            return new QosResult()
+            {
+                ErrorCode = (int) resultCode,
+                RegionResults = results,
+                ErrorMessage = errorMessage
+            };
+        }
+        
+        
+        private async Task PingWorker(RegionPinger[] regionPingers, IProducerConsumerCollection<int> seeds)
+        {
+            while(seeds.TryTake(out int seed))
+            {
+                for (int i = 0; i < regionPingers.Length; i++)
+                {
+                    int index = (i + seed) % regionPingers.Length;
```

Comment:
Added a comment. Will discuss to see if it can be better.

---
In reply to: [376138102](https://github.com/PlayFab/CSharpSDK/pull/111#discussion_r376138102) [](ancestors = 376138102)